src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
+            data = expandedData();
+            System.arraycopy(tempData, 0, data, 0, tempData.length);
+        }
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size == data.length) {
+            tempData = data;
+            data = expandedData();
+            System.arraycopy(tempData, 0, data, 0, tempData.length);
+        }
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't correctly add element by index "
+                    + index + "!");
+        }
+        System.arraycopy(data, index, data, index + 1, data.length - index - 1);
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't correctly get element by index "
+                    + index + "!");
+        }
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't correctly replace element by index "
+                    + index + "!");
+        }
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't correctly delete element by index "
+                    + index + "!");
+        }
+        temp = data[index];
+        if (index == 0) {
+            System.arraycopy(data, 1, data, 0, size);
+            size--;
+            return temp;
+        }
+        if (size == data.length) {
+            System.arraycopy(data, index + 1, data, index, data.length - 1 - index);
+            size--;
+            data[size] = null;
+            return temp;
+        }
+        if (index < size) {
+            System.arraycopy(data, index + 1, data, index, size - index);
+            size--;
+            data[size] = null;
+        }
+        return temp;
     }
 
     @Override
     public T remove(T element) {
+        if (element == null) {
+            size--;
+            return null;
+        }
+        for (int i = 0; i < size; i++) {
+            if (data[i] != null && data[i].equals(element)) {
+                temp = data[i];
+                System.arraycopy(data, i + 1, data, i, size - 1 - i);
+                size--;
+                data[size] = null;
+                return temp;
+            }
+        }
+        for (T t : data) {
+            if (t != null && !t.equals(element)) {
+                throw new NoSuchElementException("Item not found!");
+            }
+        }
         return null;
```

Comment:
why are you iterating 2 times

lets leave the first loop and remove the second loop
in the first loop lets reuse method remove(index)